testsuite/tests/typing-poly/poly.ml

```diff
@@ -1025,8 +1026,7 @@ and  ('a2, 'b2) ty2 = 'b2 -> unit constraint 'b2 = [> `V2 of ('a2, 'b2) ty1 as '
 Line 1, characters 0-83:
 1 | type ('a1, 'b1) ty1 = 'a1 -> unit constraint 'a1 = [> `V1 of ('a1, 'b1) ty2 as 'b1]
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-Error: The definition of ty1 contains a cycle:
-       [> `V1 of ('a, 'b) ty2 as 'b ] as 'a
+Error: The definition of ty1 contains a cycle
```

Comment:
This example gives an arguably slightly worse message than before. (Other cases are clear improvements, in my opinion.) We could get the best of both world by keeping the "last type informed in the cycle" parameter and printing it only in case the expansion sequence is omitted (as in this case) because it is too short. But we would also get the code complexity of both worlds, and I think that it is not worth it just to get a slight improvement in those pathological examples involving constraints.

Reply:
In the standard cases, where a cycle can be easily exhibited, this PR is a clear improvement.
On the other hand, in some harder cases where the cause is more subtle, there is no output at all.
To me this looks like it may make debugging of the compiler harder, since we cannot even point where the problem happens.
I didn't look much at the code, but would it possible to print a shorter trace, if length is the problem?

Reply:
Thanks for your feedback. I will try to tweak the PR to always show at least this "final" type.

Reply:
The new output on this example, after I rebased the PR to also trace subterm relations:

```
Error: The definition of ty1 contains a cycle:
         [> `V1 of ('a, 'b) ty2 as 'b ] as 'a contains 'b
```